model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -164,22 +173,46 @@ message ProcessBundleDescriptor {
   // (Required) A map from pipeline-scoped id to PTransform.
   map<string, org.apache.beam.model.pipeline.v1.PTransform> transforms = 2;
 
+  // Specifies that the given elements should be passed to the given transforms
+  // to kick off the bundle. Can be empty - then some of the transforms will
+  // have to start themselves without any input (e.g. the impulse transform or
+  // the read-from-gRPC transform).
+  repeated Application roots = 3;
+
   // (Required) A map from pipeline-scoped id to PCollection.
-  map<string, org.apache.beam.model.pipeline.v1.PCollection> pcollections = 3;
+  map<string, org.apache.beam.model.pipeline.v1.PCollection> pcollections = 4;
 
   // (Required) A map from pipeline-scoped id to WindowingStrategy.
-  map<string, org.apache.beam.model.pipeline.v1.WindowingStrategy> windowing_strategies = 4;
+  map<string, org.apache.beam.model.pipeline.v1.WindowingStrategy> windowing_strategies = 5;
 
   // (Required) A map from pipeline-scoped id to Coder.
-  map<string, org.apache.beam.model.pipeline.v1.Coder> coders = 5;
+  map<string, org.apache.beam.model.pipeline.v1.Coder> coders = 6;
 
   // (Required) A map from pipeline-scoped id to Environment.
-  map<string, org.apache.beam.model.pipeline.v1.Environment> environments = 6;
+  map<string, org.apache.beam.model.pipeline.v1.Environment> environments = 7;
 
   // A descriptor describing the end point to use for State API
   // calls. Required if the Runner intends to send remote references over the
   // data plane or if any of the transforms rely on user state or side inputs.
-  org.apache.beam.model.pipeline.v1.ApiServiceDescriptor state_api_service_descriptor = 7;
+  org.apache.beam.model.pipeline.v1.ApiServiceDescriptor state_api_service_descriptor = 8;
+}
+
+// Represents a partition of the bundle into two bundles: a "primary" and
+// a "residual", with the following properties:
+// - The work in primary and residual doesn't overlap, and combined, adds up
+//   to the work in the current bundle if the split hadn't happened.
+// - The current bundle, if it keeps executing, will have done none of the
+//   work under residual roots.
+// - The current bundle, if no further splits happen, will have done exactly
+//   the work under primary_roots.
+// For more rigorous definitions see https://s.apache.org/beam-breaking-fusion
+message BundleSplit {
```

Comment:
Is it worth providing an estimate of the ratio of primary to residual work?

Reply:
Yup, actually there has to be more stuff describing how we think these Application's will go, e.g. watermarks. Will think about it and add something.

Reply:
Done.